miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/file/MiniAppFileDownloader.kt

```diff
@@ -0,0 +1,128 @@
+package com.rakuten.tech.mobile.miniapp.file
+
+import android.app.Activity
+import android.content.Intent
+import android.net.Uri
+import android.util.Log
+import android.webkit.MimeTypeMap
+import androidx.annotation.VisibleForTesting
+import com.rakuten.tech.mobile.miniapp.errors.MiniAppDownloadFileError
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.launch
+import okhttp3.OkHttpClient
+import okhttp3.Request
+
+/**
+ * The file downloader of a miniapp with `onStartFileDownload` function.
+ * To start file download on the device.
+ **/
+interface MiniAppFileDownloader {
+    /**
+     * For downloading the files which has been invoked by miniapp.
+     * @param fileName return the name of the file.
+     * @param url return the download url of the file.
+     * @param headers returns the header of the file if there is any.
+     * @param onDownloadSuccess contains file name send from host app to miniapp.
+     * @param onDownloadFailed contains custom error message send from host app.
+     **/
+    fun onStartFileDownload(
+        fileName: String,
+        url: String,
+        headers: Map<String, String>,
+        onDownloadSuccess: (String) -> Unit,
+        onDownloadFailed: (MiniAppDownloadFileError) -> Unit
+    )
+}
+
+/**
+ * The default file downloader of a miniapp.
+ * @param requestCode of file downloading using an intent inside sdk, which will also be used
+ * to retrieve the Uri of the file by [Activity.onActivityResult] in the HostApp.
+ **/
+class MiniAppFileDownloaderDefault(var activity: Activity, var requestCode: Int) : MiniAppFileDownloader {
+    private lateinit var fileName: String
+    private lateinit var url: String
+    private lateinit var headers: Map<String, String>
+    private lateinit var onDownloadSuccess: (String) -> Unit
+    private lateinit var onDownloadFailed: (MiniAppDownloadFileError) -> Unit
+    private val scope: CoroutineScope = CoroutineScope(Dispatchers.IO)
+    private var okHttpClient: OkHttpClient? = null
+
+    /**
+     * Retrieve the Uri of the file by [Activity.onActivityResult] in the HostApp.
+     **/
+    fun onReceivedResult(destinationUri: Uri) {
+        val client = okHttpClient ?: OkHttpClient.Builder().build().apply { okHttpClient = this }
+        var request = createRequest(url, headers)
+        scope.launch {
+            startDownloading(destinationUri, client, request)
+        }
+    }
+
+    private fun startDownloading(destinationUri: Uri, client: OkHttpClient, request: Request) {
+        val response = client.newCall(request).execute()
+        if (response.isSuccessful) {
+            response.body?.let { responseBody ->
+                activity.contentResolver.openOutputStream(destinationUri)?.let { outputStream ->
+                    outputStream.write(responseBody.bytes())
```

Comment:
Sorry, I used `ResponseBody.bytes` in the example, but I'm not sure if that's the most efficient way to handle large files. I think it can run into out of memory issues.

Maybe it's better to use an `InputStream`? Kotlin has a convenience function for this:
```
responseBody.byteStream().copyTo(outputStream)
```